internal/controller/nimbuild_controller.go

```diff
@@ -0,0 +1,1035 @@
+		scheme:   scheme,
+		log:      log,
+		Platform: platform,
+	}
+}
+
+// +kubebuilder:rbac:groups=apps.nvidia.com,resources=nimbuilds,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps.nvidia.com,resources=nimbuilds/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=apps.nvidia.com,resources=nimbuilds/finalizers,verbs=update
+// +kubebuilder:rbac:groups=security.openshift.io,resources=securitycontextconstraints,verbs=use,resourceNames=nonroot
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch;create;delete
+// +kubebuilder:rbac:groups="",resources=pods/log,verbs=get
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
+func (r *NIMBuildReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	var err error
+	var result reconcile.Result
+
+	// Fetch the NIMBuild instance
+	nimBuild := &appsv1alpha1.NIMBuild{}
+	if err = r.Get(ctx, req.NamespacedName, nimBuild); err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "unable to fetch NIMBuild", "name", req.NamespacedName)
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	logger.Info("Reconciling", "NIMBuild", nimBuild.Name)
+	previousStatusState := nimBuild.Status.State
+
+	defer func() {
+		if err != nil {
+			r.GetEventRecorder().Eventf(nimBuild, corev1.EventTypeWarning, "ReconcileFailed",
+				"NIMBuild %s reconcile failed, msg: %s", nimBuild.Name, err.Error())
+		} else if previousStatusState != nimBuild.Status.State {
+			r.GetEventRecorder().Eventf(nimBuild, corev1.EventTypeNormal, nimBuild.Status.State,
+				"NIMBuild %s reconcile success, new state: %s", nimBuild.Name, nimBuild.Status.State)
+		}
+	}()
+	// Check if the instance is marked for deletion
+	if nimBuild.DeletionTimestamp.IsZero() {
+		// Add finalizer if not present
+		if !controllerutil.ContainsFinalizer(nimBuild, NIMBuildFinalizer) {
+			controllerutil.AddFinalizer(nimBuild, NIMBuildFinalizer)
+			if err = r.Update(ctx, nimBuild); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The instance is being deleted
+		if controllerutil.ContainsFinalizer(nimBuild, NIMBuildFinalizer) {
+			// Perform cleanup of resources
+			if err = r.cleanupNIMBuild(ctx, nimBuild); err != nil {
+				return ctrl.Result{}, err
+			}
+
+			// Remove finalizer to allow for deletion
+			controllerutil.RemoveFinalizer(nimBuild, NIMBuildFinalizer)
+			if err := r.Update(ctx, nimBuild); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
```

Comment:
```suggestion
                }
		return ctrl.Result{}, nil
```
You're missing the return to prevent reconciling deleting obj